com.intellij.diagnostic.PluginException: Element: class nl.hannahsten.texifyidea.file.LatexFile #Latex  because: different providers: com.intellij.psi.SingleRootFileViewProvider{vFile=file:///home/latex/aussehen/hbrs-inf.sty, vFileId=369, content=VirtualFileContent{size=2020}, eventSystemEnabled=true}(495d9309); com.intellij.psi.SingleRootFileViewProvider{vFile=file:///home/latex/hbrs-inf.sty, vFileId=369, content=VirtualFileContent{size=2020}, eventSystemEnabled=true}(3d528c67)

### Type of JetBrains IDE (IntelliJ, PyCharm, etc.) and version
IntelliJ IDEA 2022.1.2 (build IU-221.5787.30)

### Operating System 
Linux 5.18.1-zen1-1-zen (amd64)

### TeXiFy IDEA version
0.7.18

### Description
Happened while updating https://plugins.jetbrains.com/plugin/11070-unit-file-support-systemd-
It doesn't need a IDE restart for updates (hot reloading).
Now it happens on every file save until I restart the IDE.


### Stacktrace
```
com.intellij.diagnostic.PluginException: Element: class nl.hannahsten.texifyidea.file.LatexFile #Latex  because: different providers: com.intellij.psi.SingleRootFileViewProvider{vFile=file:///home/latex/aussehen/hbrs-inf.sty, vFileId=369, content=VirtualFileContent{size=2020}, eventSystemEnabled=true}(495d9309); com.intellij.psi.SingleRootFileViewProvider{vFile=file:///home/latex/aussehen/hbrs-inf.sty, vFileId=369, content=VirtualFileContent{size=2020}, eventSystemEnabled=true}(3d528c67)
invalidated at: see attachment [Plugin: nl.rubensten.texifyidea]
	at com.intellij.diagnostic.PluginProblemReporterImpl.createPluginExceptionByClass(PluginProblemReporterImpl.java:23)
	at com.intellij.diagnostic.PluginException.createByClass(PluginException.java:92)
	at com.intellij.psi.util.PsiUtilCore.ensureValid(PsiUtilCore.java:478)
	at com.intellij.psi.impl.source.PsiFileImpl.loadTreeElement(PsiFileImpl.java:192)
	at com.intellij.psi.impl.source.PsiFileImpl.calcTreeElement(PsiFileImpl.java:734)
	at com.intellij.psi.impl.source.SpineRef.getNode(SpineRef.java:26)
	at com.intellij.extapi.psi.StubBasedPsiElementBase.getNode(StubBasedPsiElementBase.java:134)
	at com.intellij.extapi.psi.ASTDelegatePsiElement.getFirstChild(ASTDelegatePsiElement.java:99)
	at com.intellij.psi.util.PsiTreeUtil.getChildrenOfTypeAsList(PsiTreeUtil.java:412)
	at nl.hannahsten.texifyidea.psi.impl.LatexCommandsImpl.getParameterList(LatexCommandsImpl.java:46)
	at nl.hannahsten.texifyidea.util.PsiCommandsKt.requiredParameters(PsiCommands.kt:197)
	at nl.hannahsten.texifyidea.psi.LatexCommandsImplUtilKt.readFirstParam$lambda-3(LatexCommandsImplUtil.kt:128)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:920)
	at nl.hannahsten.texifyidea.psi.LatexCommandsImplUtilKt.readFirstParam(LatexCommandsImplUtil.kt:126)
	at nl.hannahsten.texifyidea.psi.LatexCommandsImplUtilKt.getReferences(LatexCommandsImplUtil.kt:30)
	at nl.hannahsten.texifyidea.psi.LatexPsiImplUtil.getReferences(LatexPsiImplUtil.java:26)
	at nl.hannahsten.texifyidea.psi.impl.LatexCommandsImpl.getReferences(LatexCommandsImpl.java:57)
	at nl.hannahsten.texifyidea.util.files.ImportPackageKt.searchFileByImportPaths(ImportPackage.kt:19)
	at nl.hannahsten.texifyidea.inspections.latex.probablebugs.LatexInclusionLoopInspection.inspectFile(LatexInclusionLoopInspection.kt:42)
	at nl.hannahsten.texifyidea.inspections.TexifyInspectionBase.checkFile(TexifyInspectionBase.kt:137)
	at com.intellij.codeInspection.LocalInspectionTool$1.visitFile(LocalInspectionTool.java:141)
	at com.intellij.extapi.psi.PsiFileBase.accept(PsiFileBase.java:60)
	at com.intellij.codeInsight.daemon.impl.InspectionRunner.lambda$processInOrder$10(InspectionRunner.java:317)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1152)
	at com.intellij.codeInsight.daemon.impl.InspectionRunner.lambda$processInOrder$11(InspectionRunner.java:296)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.executeByImpatientReader(ReadMostlyRWLock.java:174)
	at com.intellij.openapi.application.impl.ApplicationImpl.executeByImpatientReader(ApplicationImpl.java:213)
	at com.intellij.codeInsight.daemon.impl.InspectionRunner.lambda$processInOrder$12(InspectionRunner.java:330)
	at com.intellij.util.AstLoadingFilter.forceAllowTreeLoading(AstLoadingFilter.java:159)
	at com.intellij.util.AstLoadingFilter.forceAllowTreeLoading(AstLoadingFilter.java:151)
	at com.intellij.codeInsight.daemon.impl.InspectionRunner.lambda$processInOrder$13(InspectionRunner.java:294)
	at com.intellij.util.AstLoadingFilter.disallowTreeLoading(AstLoadingFilter.java:130)
	at com.intellij.util.AstLoadingFilter.disallowTreeLoading(AstLoadingFilter.java:119)
	at com.intellij.codeInsight.daemon.impl.InspectionRunner.lambda$processInOrder$14(InspectionRunner.java:294)
	at com.intellij.concurrency.JobLauncherImpl$1MyProcessQueueTask.lambda$call$0(JobLauncherImpl.java:297)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$12(CoreProgressManager.java:608)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:683)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:639)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:607)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.concurrency.JobLauncherImpl$1MyProcessQueueTask.call(JobLauncherImpl.java:282)
	at com.intellij.concurrency.JobLauncherImpl$1MyProcessQueueTask.call(JobLauncherImpl.java:270)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(ForkJoinTask.java:1448)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: com.intellij.psi.PsiInvalidElementAccessException: Element: class nl.hannahsten.texifyidea.file.LatexFile #Latex  because: different providers: com.intellij.psi.SingleRootFileViewProvider{vFile=file:///home/latex/aussehen/hbrs-inf.sty, vFileId=369, content=VirtualFileContent{size=2020}, eventSystemEnabled=true}(495d9309); com.intellij.psi.SingleRootFileViewProvider{vFile=file:///home/martin/Sync/Studium/HBRS_SA/Semester_10/Praxisprojekt/HBRS-
```

--- a/src/nl/hannahsten/texifyidea/index/IndexUtilBase.kt
+++ b/src/nl/hannahsten/texifyidea/index/IndexUtilBase.kt
@@ -7,9 +7,11 @@ import com.intellij.openapi.project.DumbService
 import com.intellij.openapi.project.Project
 import com.intellij.psi.PsiElement
 import com.intellij.psi.PsiFile
+import com.intellij.psi.SmartPsiElementPointer
 import com.intellij.psi.search.GlobalSearchScope
 import com.intellij.psi.stubs.StubIndex
 import com.intellij.psi.stubs.StubIndexKey
+import com.intellij.refactoring.suggested.createSmartPointer
 import nl.hannahsten.texifyidea.util.files.documentClassFileInProject
 import nl.hannahsten.texifyidea.util.files.findRootFile
 import nl.hannahsten.texifyidea.util.files.referencedFileSet
@@ -31,7 +33,7 @@ abstract class IndexUtilBase<T : PsiElement>(
 ) {
 
     /** Cache the index items to avoid unnecessary get actions from the index, which take a long time (50-100ms) even for a small index, which can be problematic if index is accessed many times per second. */
-    var cache: MutableMap<Project, MutableMap<GlobalSearchScope, Collection<T>>> = mutableMapOf()
+    var cache: MutableMap<Project, MutableMap<GlobalSearchScope, Collection<SmartPsiElementPointer<T>>>> = mutableMapOf()
 
     /**
      * Get all the items in the index in the given file set.
@@ -116,10 +118,10 @@ abstract class IndexUtilBase<T : PsiElement>(
      */
     fun getItems(project: Project, scope: GlobalSearchScope, useCache: Boolean = true): Collection<T> {
         if (useCache) {
-            cache[project]?.get(scope)?.let { return it }
+            cache[project]?.get(scope)?.let { return it.mapNotNull { pointer -> pointer.element } }
         }
         val result = getKeys(project).flatMap { getItemsByName(it, project, scope) }
-        cache.getOrPut(project) { mutableMapOf() }[scope] = result
+        cache.getOrPut(project) { mutableMapOf() }[scope] = result.map { it.createSmartPointer() }
         return result
     }
 
@@ -151,7 +153,7 @@ abstract class IndexUtilBase<T : PsiElement>(
      * @param project
      *          The project instance.
      */
-    fun getKeys(project: Project): Array<String> {
+    private fun getKeys(project: Project): Array<String> {
         return if (!DumbService.isDumb(project)) {
             StubIndex.getInstance().getAllKeys(indexKey, project).toTypedArray()
         }

--- a/src/nl/hannahsten/texifyidea/inspections/latex/codestyle/LatexFigureNotReferencedInspection.kt
+++ b/src/nl/hannahsten/texifyidea/inspections/latex/codestyle/LatexFigureNotReferencedInspection.kt
@@ -13,11 +13,14 @@ import nl.hannahsten.texifyidea.inspections.TexifyInspectionBase
 import nl.hannahsten.texifyidea.lang.magic.MagicCommentScope
 import nl.hannahsten.texifyidea.psi.LatexCommands
 import nl.hannahsten.texifyidea.psi.LatexParameterText
-import nl.hannahsten.texifyidea.util.*
 import nl.hannahsten.texifyidea.util.files.commandsInFileSet
+import nl.hannahsten.texifyidea.util.firstChildOfType
+import nl.hannahsten.texifyidea.util.firstParentOfType
+import nl.hannahsten.texifyidea.util.isFigureLabel
 import nl.hannahsten.texifyidea.util.labels.findLabelingCommandsInFile
 import nl.hannahsten.texifyidea.util.labels.getLabelReferenceCommands
 import nl.hannahsten.texifyidea.util.magic.CommandMagic
+import nl.hannahsten.texifyidea.util.requiredParameter
 import java.util.*
 
 open class LatexFigureNotReferencedInspection : TexifyInspectionBase() {
@@ -47,7 +50,7 @@ open class LatexFigureNotReferencedInspection : TexifyInspectionBase() {
         val referenceCommands = file.project.getLabelReferenceCommands()
         for (command in file.commandsInFileSet()) {
             // Don't resolve references in command definitions
-            if (command.parent.firstParentOfType(LatexCommands::class)?.name in CommandMagic.commandDefinitionsAndRedefinitions ||
+            if (command.parent?.firstParentOfType(LatexCommands::class)?.name in CommandMagic.commandDefinitionsAndRedefinitions ||
                 referenceCommands.contains(command.name)
             ) {
                 command.referencedLabelNames.forEach { figureLabels.remove(it) }

--- a/src/nl/hannahsten/texifyidea/util/Packages.kt
+++ b/src/nl/hannahsten/texifyidea/util/Packages.kt
@@ -222,7 +222,7 @@ object PackageUtils {
 
             // Just skip conditionally included packages, because it is too expensive to determine whether
             // they are really included or not
-            if (cmd.parent.firstParentOfType(LatexCommands::class)?.name == "\\" + LatexGenericRegularCommand.ONLYIFSTANDALONE.command) {
+            if (cmd.parent?.firstParentOfType(LatexCommands::class)?.name == "\\" + LatexGenericRegularCommand.ONLYIFSTANDALONE.command) {
                 continue
             }
 

